src/Identity/IdentityServer/BaseRequestValidator.cs

```diff
@@ -133,6 +142,15 @@ public abstract class BaseRequestValidator<T> where T : class
             var verified = await VerifyTwoFactor(user, twoFactorOrganization,
                 twoFactorProviderType, twoFactorToken);
 
+            var isOtpCached = Core.Utilities.DistributedCacheExtensions.TryGetValue(_distributedCache, user.Email, out string cachedToken);
```

Comment:
❓ The task mentions considering the provider as well. Is that still needed?

Reply:
I don't think so. At its core this solution is to just track a user and the TOTPs they've used. I'm not sure why we care which provider it was. The possibility of a collision between TOTP providers I feel like is slim to none when considering most users probably use a single TOTP provider and the cache is cleared regularly.

If there is an argument for tracking the provider then I'm all for implementing it, I just didn't see the need, but I could also be missing something.